### PR TITLE
gh-125543: Add an internal C API for dynamic arrays

### PR DESCRIPTION
**Pull request by ZeroIntensity**
Co-authored-by: Kirill Podoprigora <kirill.bast9@mail.ru>

* Issue: gh-125543

### CODE REVIEW

File: Include/internal/pycore_dynarray.h [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+#define _PyDynArray_DEFAULT_SIZE 16
+
+typedef void (*_PyDynArray_Deallocator)(void *);
+
+typedef struct {
+    void **items;
+    Py_ssize_t capacity;
+    Py_ssize_t length;
+    _PyDynArray_Deallocator deallocator;
```
Comment (Eclips4) [inline Include/internal/pycore_dynarray.h:44]: How about taking an idea from `Py_hastable_t`?

It has a `Py_hashtable_allocator_t alloc` field which can be used both for allocation and deallocation.

File: Python/dynarray.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        }
+    }
+    PyMem_RawFree(array->items);
+
+    // It would be nice if others could reuse the allocation for another
+    // dynarray later, so clear all the fields.
+    array->items = NULL;
+    array->length = 0;
+    array->capacity = 0;
+    array->deallocator = NULL;
```
Comment (Eclips4) [inline Python/dynarray.c:149]: Hm, setting the `deallocator` to a `NULL` looks weird to me. I can't say that this doesn't have a use case, but I guess this isn't what people expect from something called `...clear`.

Comment (ZeroIntensity) [reply Python/dynarray.c:149]: The idea of `Clear` here is to completely reset the hash table, not remove all the values. It's sort of like a deallocator, except it doesn't free the underling pointer.

File: Include/internal/pycore_dynarray.h (hunk 1 above)
Comment (ZeroIntensity) [reply Include/internal/pycore_dynarray.h:44]: I decided that's too inconvenient for users. In basically all cases, there's no need to customize the allocator. We're using the `Raw` domain anyway, so it's not like we have to worry about the caller holding the GIL.

File: Python/dynarray.c [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    if (array->length == array->capacity)
+    {
+        // Need to resize
+        array->capacity *= 2;
+        void **new_items = PyMem_RawRealloc(
+                                            array->items,
+                                            sizeof(void *) * array->capacity);
+        if (new_items == NULL)
+        {
+            --array->length;
```
Comment (Eclips4) [inline Python/dynarray.c]: Can we add a comment why this(line 50) is needed?

File: Python/dynarray.c [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            return -1;
+        }
+
+        // XXX Zero-out the new capacity?
```
Comment (Eclips4) [inline Python/dynarray.c]: Why is this needed?

Comment (ZeroIntensity) [reply Python/dynarray.c]: `realloc()` doesn't initialize memory to 0, so there's junk memory in `items`. However, it's not possible for someone to touch that junk memory unless they're manually using `items`, so it's not an issue. I can remove the comment if you want.

File: Python/dynarray.c (hunk 3 above)
Comment (ZeroIntensity) [reply Python/dynarray.c]: I'll move that, it's better suited for the caller.

File: Python/dynarray.c (hunk 4 above)
Comment (Eclips4) [reply Python/dynarray.c]: I think that if someone directly touches the `items` field and something goes wrong, it's their fault. So, I'm in favor of removing this comment.

**Review by Eclips4 — APPROVED** [review]
The current implementation and API look good to me. Let's see what others think.

**Review by vstinner — CHANGES_REQUESTED** [review]
At this point, I'm not convinced that there is a need for such API. I suggest to discuss it on the issue.
